solver/jobs.go

```diff
@@ -282,7 +287,12 @@ func (jl *Solver) setEdge(e Edge, newEdge *edge) {
 		return
 	}
 
-	st.setEdge(e.Index, newEdge)
+	targetSt, ok := jl.actives[targetEdge.edge.Vertex.Digest()]
+	if !ok {
+		return
+	}
```

Comment:
Can we retrieve the state directly from the target edge? The edge itself contains a reference to `activeOp` and this is always an instance of `sharedOp`. `sharedOp` contains a reference to the state that I believe should be identical to this.

It might be better if we add a method to the `activeOp` interface like this:

```
func State() *state
```

Then have `sharedOp` implement that to return the state. These are all private interfaces so it shouldn't be a problem.

I would also say I'm a bit uncomfortable with the idea of returning when this fails. It feels to me like we should just pass `nil` to `setEdge` and have `setEdge` check if the state is nil. While I don't think anything could ever cause line 292 to trigger, if it does, I'd like it to just continue on through the old control-flow for safety reasons.

Reply:
It is somewhat intentional that this can't be directly accessed. We want to avoid that the state machine in scheduler has capability of directly modifying the builder state or the shared graph. This is why there is the `edgeFactory` and `activeOp` interface and the scheduler always needs to go through these methods.

> I would also say I'm a bit uncomfortable with the idea of returning when this fails. 

This indeed isn't ideal, but it is not new to this PR. It could error, but that is bit tricky from the event loop and should be separate PR. But the main issue is that if this returns early it is basically some reference counting bug and graph is in an inconsistent state. So even if we error, it would be hard to fix up this state and probably something else would fail. We can't just skip the merge because this function is only called after the merge has already completed as a final pointer switch.

Reply:
I think I misunderstood your second comment and you just suggested skipping setting progress writer if this new state lookup fails for any reason. Yes, I can change that.